File: rust/examples/test_logger.rs
use std::sync::Arc;
use microservice_toolbox::utils::logger::{UniLogger, Logger};
use microservice_toolbox::conn_manager::manager::new_network_manager_with_logger;
use microservice_toolbox::config::loader::load_config_with_logger;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!(">>> Initializing UniLogger (compiled Go engine)...");
    
    // 1. Initialize UniLogger (this loads the DLL via unilog crate)
    let config_profile = "standalone";
    let u_log = unilog_rs::UniLog::new("standalone", "test-rust", "standard", unilog_rs::LogLevel::Info, true)?;
    let logger: Arc<dyn Logger> = Arc::new(UniLogger::new(u_log));

    logger.info("Rust Toolbox modernized with compiled universal-logger!");

    // 2. Test NetworkManager with UniLogger
    println!(">>> Testing NetworkManager with UniLogger...");
    let nm = new_network_manager_with_logger(
        2,      // max retries
        200,    // base delay
        1000,   // max delay
        1000,   // timeout
        2.0,    // backoff
        0.1,    // jitter
        Some(logger.clone()),
    );

    // This should fail to connect but LOG via our UniLogger
    println!(">>> Expected failure logs should appear below (from Go core):");
    let _ = nm.connect_with_retry("localhost".to_string(), "1234".to_string()).await;

    // 3. Test AppConfig with UniLogger
    println!(">>> Testing AppConfig with UniLogger...");
    let config = load_config_with_logger(config_profile, Some(logger.clone()))?;
    println!(">>> Config loaded for profile: {}", config.profile);

    println!(">>> SUCCESS: Rust Toolbox modernized with compiled universal-logger!");
    Ok(())
}

File: python/test_logger.py
import os
import sys
from pathlib import Path

# Add universal-logger/python to sys.path to find 'unilog' package
uni_logger_path = str(Path(__file__).parent.parent.parent / "universal-logger" / "python")
sys.path.append(uni_logger_path)

try:
    from unilog.facade import UniLog

    from microservice_toolbox.config.loader import load_config_with_logger
    from microservice_toolbox.conn_manager.manager import new_network_manager_with_logger

    print(">>> Initializing UniLog (compiled)...")
    native_log = UniLog(config_profile="standalone", app_name="test-python")
    logger = native_log

    print(">>> Testing NetworkManager with UniLogger...")
    nm = new_network_manager_with_logger(max_retries=2, logger=logger)
    try:
        # This will fail and log retries
        nm.connect_with_retry("localhost", "1234", "127.0.0.1", "test")
    except Exception as e:
        print(f">>> Expected failure: {e}")

    print(">>> Testing AppConfig with UniLogger...")
    # Requires a standalone.yaml to exist in current dir
    if os.path.exists("standalone.yaml"):
        conf = load_config_with_logger("standalone", logger=logger)
        print(f">>> Config loaded for profile: {conf.profile}")
    else:
        print(">>> Skipping AppConfig test (standalone.yaml not found)")

    print(">>> SUCCESS: Python Toolbox modernized with compiled universal-logger!")

except ImportError as e:
    print(f">>> FAILED: Could not import unilog or toolbox modules: {e}")
except Exception as e:
    print(f">>> FAILED: Unexpected error: {e}")
    import traceback

    traceback.print_exc()
